### PR TITLE
[RFC] ttv: recommended project structure 1 - hacker news

### PR DESCRIPTION
**Pull request by yuhan**
### Summary & Motivation

Goals:
- Don't make me think: users should be able to quickly iterate (e.g. add code) without thinking much about where it should be.
- Findability: users should be able to quickly find the code regardless of their familiarity to the business logics.
- Flexibility to grow: users should be able to organize at their own pace when they feel things have grown too big. (Don't over optimize too early)

Note that we haven't decided if the hacker news asset example would be the "ultimate" go-to example that we point users to, but since this is the most complex project, let's start with this to unfold the difficult patterns.

Proposed changes:
- Consolidate where we define schedules and sensors. 
  * Concretely, this RFC moves all asset schedules and sensors to repository.py; they were separate in `activity_analytics/__init__.py`, `core/__init__.py`, `recommender/__init__.py`. I think it was hard to find and get an overview of all sensor/schedule defs in a repo.
- Put all assets under `assets/` dir, separated by business domain inside the `assets/` dir
  * It was separate in `activity_analytics/assets`, `core/assets`, `recommender/assets`. Having the business domains to be top-level dirs in a dagster project module makes it hard to discover assets in a project for new authors/collaborators.
- `partitions/__init__.py` -> `partitions.py`

Check out project layout here: https://github.com/dagster-io/dagster/tree/yuhan/ttv-project-struct-1/examples/hacker_news_assets

### CODE REVIEW

**Comment by yuhan** [comment]
Current dependencies on/for this PR:
* master
  * **PR #8936** <a href="https://app.graphite.dev/github/pr/dagster-io/dagster/8936" target="_blank">Graphite</a>  👈

This comment was auto-generated by [Graphite](https://app.graphite.dev/github/pr/dagster-io/dagster/8936?utm_source=stack-comment).

File: examples/hacker_news_assets/hacker_news_assets/assets/__init__.py [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+
+from . import activity_analytics, core, recommender
+
+# TBD: is it a good practice to build asset groups in assets/__init__.py?
+# * pros:
+#       * users can quickly find code even if they are not familiar with the given project
+#       * users can quickly add asset group without thinking about schedules and sensors
+#       * users can easily build asset groups that are cross business domains
+# * cons:
+#       * users will likely need to think about schedules and sensors separately from building assets
```
Comment (yuhan) [inline examples/hacker_news_assets/hacker_news_assets/assets/__init__.py:16]: RFC

File: examples/hacker_news_assets/hacker_news_assets/repository.py [hunk 2]
```diff
@@ -0,0 +1,80 @@
...
+from hacker_news_assets.partitions import hourly_partitions
+
+
+# TBD: is it a good practice to put all sensors and sensors live inside repository?
+# * pros:
+#       * no extra schedule and sensor layers at starter - sensors and schedules are mostly just wrappers around jobs or asset groups
+#       * users can choose to separate them out once the project grows
+# * cons:
+#       * it'd be a lot of code "in your face" as the the repository is the main entry of a dagster
+#         project.
```
Comment (yuhan) [inline examples/hacker_news_assets/hacker_news_assets/repository.py:32]: RFC

**Review by sryza — COMMENTED** [review]
This looks pretty nice!

File: examples/hacker_news_assets/hacker_news_assets/repository.py (hunk 2 above)
Comment (clairelin135) [reply examples/hacker_news_assets/hacker_news_assets/repository.py]: I am in support of sensors and schedules living inside `repository.py`. I think they should be defined outside of the `@repository` function but within the file (like you did in this PR) to make the contents of the repository readable.

I think the inconvenience of having to do multiple imports if schedules/sensors existed in their own file outweighs the "code in your face"

File: examples/hacker_news_assets/hacker_news_assets/assets/__init__.py (hunk 1 above)
Comment (clairelin135) [reply examples/hacker_news_assets/hacker_news_assets/assets/__init__.py]: I think it feels a little hidden to put asset code within an __init__ file (I personally wouldn't look there for code) but I support placing assets in their own `assets.py` file for the reasons you pointed out

File: examples/hacker_news_assets/src/hacker_news_assets/repository.py [hunk 3]
```diff
@@ -0,0 +1,78 @@
...
+    repository,
+    with_resources,
+)
+
+# TBD: is it a good practice to put all sensors and sensors live inside repository?
+# * pros:
+#       * no extra schedule and sensor layers at starter - sensors and schedules are mostly just wrappers around jobs or asset groups
+#       * users can choose to separate them out once the project grows
+# * cons:
+#       * it'd be a lot of code "in your face" as the the repository is the main entry of a dagster
```
Comment (sryza) [inline examples/hacker_news_assets/src/hacker_news_assets/repository.py:29]: what would the alternative be?

**Comment by yuhan** [comment]
moved to https://github.com/dagster-io/dagster/pull/9191

**Pull request closed without merging**